### PR TITLE
planner: allow TableDual to match the required property if the row count is 0

### LINKED ISSUE #16841: Can't find a proper physical plan for this query
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
``` sql
drop table if exists t1, t2, t3;
create table t1 (a int, b tinyint)  partition by range (a) (
    partition p0 values less than (10),
    partition p1 values less than (20),
    partition p2 values less than (30),
    partition p3 values less than (40),
    partition p4 values less than MAXVALUE
);
insert into t1 values(NULL, NULL), (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (10, 10), (11, 11), (12, 12), (13, 13), (14, 14), (15, 15), (20, 20), (21, 21), (22, 22), (23, 23), (24, 24), (25, 25), (30, 30), (31, 31), (32, 32), (33, 33), (34, 34), (35, 35), (36, 36), (40, 40), (50, 50), (80, 80), (90, 90), (100, 100);

create table t2 (a int, b bigint) partition by hash(a) partitions 10;
insert into t2 values (NULL, NULL), (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (8, 8), (9, 9), (10, 10), (11, 11), (12, 12), (13, 13), (14, 14), (15, 15), (16, 16), (17, 17), (18, 18), (19, 19), (20, 20), (21, 21), (22, 22), (23, 23);

create table t3 (a int, b mediumint);
insert into t3 values (NULL, NULL), (0, 0), (1, 1), (2, 2), (3, 3), (4, 4), (5, 5), (6, 6), (7, 7), (8, 8), (9, 9), (10, 10), (11, 11), (12, 12), (13, 13), (14, 14), (15, 15), (16, 16), (17, 17), (18, 18), (19, 19), (20, 20), (21, 21), (22, 22), (23, 23);

select /*+ MERGE_JOIN(t1, t3) */ * from t3 left join t1 partition (p0) on t3.a = t1.a where t1.a > 100 order by t1.a, t1.b, t3.a, t3.b;
```

### 2. What did you expect to see?
Query works fine.

### 3. What did you see instead?
``` sql
tidb> select /*+ MERGE_JOIN(t1, t3) */ * from t3 left join t1 partition (p0) on t3.a = t1.a where t1.a > 100 order by t1.a, t1.b, t3.a, t3.b;
ERROR 1815 (HY000): Internal : Can't find a proper physical plan for this query
```

### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
``` sql
tidb> select tidb_version();
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                        |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v4.0.0-beta.2-343-ge607157ef
Git Commit Hash: e607157efef0f76878d18a667b233061ce43545a
Git Branch: master
UTC Build Time: 2020-04-26 07:46:48
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve?

Issue Number: close #16841  

Problem Summary:

Before this PR, PhysicalTableDual cannot match any required property(and it will return an invalidTask in findBestTask if the property is not empty). 
But the case in #16841 use a `MergeJoin` hint to enforce to use the merge join. While one child of the LogicalJoin is a LogicalTableDual, which cannot match the property. So the optimizer cannot find a physical plan for that SQL.

### What is changed and how it works?

What's Changed:

This PR allow TableDual to match the required property if the row count is 0.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

### Release note 
* Fix the scenario that the planner cannot find a physical plan when there is a TableDual.

### CODE REVIEW

File: planner/core/find_best_task.go [hunk 1]
```diff
@@ -104,7 +104,10 @@ func GetPropByOrderByItemsContainScalarFunc(items []*ByItems) (*property.Physica
 }
 
 func (p *LogicalTableDual) findBestTask(prop *property.PhysicalProperty) (task, error) {
-	if !prop.IsEmpty() {
+	// If the required property is not empty and the row count > 0,
+	// we cannot ensure this required property.
+	// But if the row count is 0, we don't need to care about the property.
```
Comment (lzmhhh123) [inline planner/core/find_best_task.go:109]: What about row count is one?

Comment (francis0407) [reply planner/core/find_best_task.go]: Good catch! Fixed. PTAL

**Pull request merged**